README.md
# Readable
[![CI](https://github.com/hinto-janai/readable/actions/workflows/ci.yml/badge.svg)](https://github.com/hinto-janai/readable/actions/workflows/ci.yml) [![crates.io](https://img.shields.io/crates/v/readable.svg)](https://crates.io/crates/readable) [![docs.rs](https://docs.rs/readable/badge.svg)](https://docs.rs/readable)

![readable](https://github.com/hinto-janai/readable/assets/101352116/2b4c0c1b-c80e-4b7a-a8b9-b86375382db9)

Human **readable** strings.

This library:
- Transforms various data types into human-readable strings
- Parses raw string data into human-readable versions
- Provides various string types and utilities

Most of the strings are implemented as fixed sized stack allocated arrays that are [`Copy`](https://doc.rust-lang.org/stable/std/marker/trait..html)-able.

In general, `readable` types are often used where you need to quickly format some data into a more human-friendly string, display it, then throw it away (although most `readable` types are perfectly fine to permanently store).

Creation of `readable` types is relatively performant.

## Examples
#### Unsigned
```rust
use readable::num::*;
assert_eq!(Unsigned::from(1000_u64), "1,000");
```
#### Int
```rust
use readable::num::*;
assert_eq!(Int::from(-1000), "-1,000");
```
#### Float
```rust
use readable::num::*;
assert_eq!(Float::from(1000.123), "1,000.123");
```
#### Percent
```rust
use readable::num::*;
assert_eq!(Percent::from(1000.123), "1,000.12%");
```
#### Runtime
```rust
use readable::run::*;
assert_eq!(Runtime::from(311.123),      "5:11");
assert_eq!(RuntimePad::from(311.123),   "00:05:11");
assert_eq!(RuntimeMilli::from(311.123), "00:05:11.123");
```
#### Uptime
```rust
use readable::up::*;
assert_eq!(Uptime::from(172799_u32),     "1d, 23h, 59m, 59s");
assert_eq!(UptimeFull::from(172799_u32), "1 day, 23 hours, 59 minutes, 59 seconds");
assert_eq!(Htop::from(172799_u32),       "1 day, 23:59:59");
```
#### Date
```rust
use readable::date::*;
assert_eq!(Date::from_ymd(2014, 12, 31).unwrap(), "2014-12-31");
assert_eq!(Nichi::new(2014, 12, 31).unwrap(),     "Wed, Dec 31, 2014");
assert_eq!(NichiFull::new(2014, 12, 31).unwrap(), "Wednesday, December 31st, 2014");
```
#### Time
```rust
use readable::time::*;
assert_eq!(Time::new(86399),     "11:59:59 PM");
assert_eq!(Military::new(86399), "23:59:59");
```
#### Byte
```rust
use readable::byte::*;
assert_eq!(Byte::from(1234), "1.234 KB");
```

## Comparison
All number types implement `PartialEq` against `str` and their internal numbers.

This is comparing `b`'s inner `String`:
```rust
use readable::up::*;
let a = std::time::Duration::from_secs(86399);
let b = UptimeFull::from(a);
assert_eq!(b, "23 hours, 59 minutes, 59 seconds");
```
This is comparing `a`'s inner `i64`:
```rust
use readable::num::*;
let a = Int::from(-1000);
assert_eq!(a, -1000);
```
This compares both the `u64` AND `String` inside `a` and `b`:
```rust
use readable::num::*;
let a = Unsigned::from(1000_u64);
let b = Unsigned::from(1000_u64);
assert_eq!(a, b);
```

## Arithmetic
All number types implement the common arithmetic operators `+`, `-`, `/`, `*`, `%`, outputting a new `Self`.
#### `+` Addition
```rust
use readable::num::*;
let f1 = Float::from(1.0);
let f2 = Float::from(2.0);
assert_eq!(f1 + f2, 3.0);
```
#### `-` Subtraction
```rust
use readable::num::*;
let p50 = Percent::from(50.0);
let p25 = Percent::from(25.0);
assert_eq!(p50 - p25, "25.00%");
```
#### `/` Division
```rust
use readable::num::*;
let u100 = Unsigned::from(100_u64);
let u10  = Unsigned::from(10_u64);
assert_eq!(u100 / u10, 10);
```
#### `*` Multiplication
```rust
use readable::num::*;
let u10 = Unsigned::from(10_u64);
assert_eq!(u10 * u10, Unsigned::from(100_u64));
```
#### `%` Modulo
```rust
use readable::num::*;
let u10 = Unsigned::from(10_u64);
assert_eq!(u10 % u10, 0);
```

## Feature Flags
These features are for (de)serialization.

| Feature Flag | Purpose |
|--------------|---------|
| `serde`      | Enables [`serde`](https://docs.rs/serde)'s `Serialize` & `Deserialize`
| `bincode`    | Enables [`bincode 2.0.0-rc.3`](https://docs.rs/bincode/2.0.0-rc.3/bincode/index.html)'s `Encode` & `Decode`
| `borsh`      | Enables [`borsh`](https://docs.rs/borsh)'s `BorshSerialize` & `BorshDeserialize`

## MSRV
The Minimum Supported Rust Version is `1.71.0`.

CHANGELOG.md
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Types of changes:
- `Added` for new features
- `Changed` for changes in existing functionality
- `Deprecated` for soon-to-be removed features
- `Removed` for now removed features
- `Fixed` for any bug fixes
- `Security` in case of vulnerabilities


---


# readable Unreleased


---


# readable v0.16.0 - 2023-12-26
## Added
- Impl `TryFrom<[u8]>` for `Str`
- Impl `borsh::{BorshSerialize, BorshDeserialize}` on most types

## Changed
- Modules are now feature gated
- Types are no longer re-exported, must be accessed through modules
- `Float` and `Percent` now use `Str` internally, allows `const`, fixed-size stack usage, and `Copy`
- `HeadTail` now requires `AsRef<str>`
- `Str::invalid()` is now `const`
- `Str::as_str()` now panics in debug mode when `Str::invalid()` returns `true`
- `Str::from_bytes_exact()` takes `impl AsRef<[u8]>` instead of `&[u8]`

## Removed
- `Str::capacity()`

## Fixed
- `Int::try_from()` a `NonZeroU*`


---


# readable v0.15.0 - 2023-12-21
## Changed
- `#[must_use]` on applicable functions
- `Str::as_bytes_mut()` is now `unsafe`
- `Str::from_bytes_exact()` is now `unsafe`
- `Str::from_str_exact()` takes `impl AsRef<str>` instead of `&str`

## Removed
- Redundant function constructors for `const` types (e.g `Uptime::day()` -> `Uptime::DAY`)


---


# readable v0.14.0 - 2023-12-03
## Added
- `readable::date::*`:
	* impl `TryFrom<(u16, u8, u8)>`
	* impl `TryFrom<(u16, u8)>`
	* impl `TryFrom<u16>`
- `readable::str::Str`:
	* impl `std::fmt::Write`
	* impl `Index`
	* impl `Extend<char>`, `Extend<&str>`
	* impl `Deref` and `AsRef`
	* impl `Add` and `AddAssign`
	* `push_{str,char}_unchecked()` -> `push_{str,char}_panic()`
	* add `push_char_*()` variants
	* add `push_str_saturating()`
	* add `remove()`
	* add `pop()`
	* add `truncate()`
	* add `make_ascii_uppercase()`, `make_ascii_lowercase()`
	* add `as_bytes_mut()`, `as_mut_ptr()`, `as_str_mut()`


---


# readable v0.13.0 - 2023-10-23
## Added
* `Time` type for clock time - `11:59:59 PM`
* `Military` type for military-style clock time - `23:59:59`
* `readable::time` free functions (`unix()`, `unix_clock()`, `secs_to_hms()`, etc)
* `readable::date` free functions (`from_unix()`, `as_unix()`, etc)
* `SysUptime`, `SysTime`, `SysDate` traits for getting live-system info

## Changed
* Uptime modules renamed for clarity `readable::time` -> `readable::up`
* Time-like types from `readable::date` into `time` module


---


# readable v0.12.0 - 2023-10-19
## Added
* `Nichi` type for calendar date formatting (`Fri, Dec 25, 2020`)
* `NichiFull` type for calendar date formatting (`Friday, December 25th, 2020`)
* `TimeUnit` type for remainder time measuring (`61` == `1 minute, 1 second`)
* `From` impl for `std::time::Duration` from all `readable::time` and `readable::run` types
* `Default` impl for `Str`

## Removed
* Feature-flags (except for `serde` & `bincode`)


---


# readable v0.11.0 - 2023-10-15
## Added
* `Htop` type in `readable::time` for [`htop`](https://github.com/htop-dev/htop)-style uptime formatting
* `Uptime` trait for getting system uptime and formatting directly into a `Time`, `TimeFull` or `Htop`
* `TryFrom` implementations from smart pointer strings (`Arc<str>`, `Box<str>`, etc) for `Str`

## Changed
* `Str::from_str()` -> `Str::from_str_exact()`
* `Str::from_bytes()` -> `Str::from_bytes_exact()`
* `Time::UNKNOWN` changed from `???` to `(unknown)`
* `TimeFull::UNKNOWN` changed from `???` to `(unknown)`


---


# readable v0.10.0 - 2023-10-04
First "public" release.


src/time.rs
//---------------------------------------------------------------------------------------------------- Use
#[cfg(feature = "serde")]
use serde::{Serialize,Deserialize};
use compact_str::{CompactString};
use std::fmt::Write;
use crate::macros::*;

//---------------------------------------------------------------------------------------------------- Time
/// Human-readable [`std::time::Duration`].
///
/// **The input is always assumed to be in seconds.**
///
/// [`Time::from`] input can be:
/// - [`u8`], [`u16`], [`u32`], [`u64`], [`usize`]
/// - [`std::time::Duration`], [`std::time::Instant`]
///
/// The lowest unit is `second`, the highest is `year`, and `week` is skipped in favor of `7 days`.
///
/// ## Inlining
/// If the feature flag `inline_time` is enabled, inputs ranging from `0..3660` into
/// [`Time::from`] will return inlined static bytes, the max string representation of this is `1 hour, 1 minute`.
///
/// **Warning:** This feature is disabled by default. While it increases speed,
/// it also _heavily_ increases build time and binary size.
///
/// ## Cloning
/// [`Clone`] may be expensive:
/// ```rust
/// # use readable::Time;
/// // Cheap (stack allocated string).
/// let a = Time::from(60_u8);
/// assert!(a == "1 minute");
/// let b = a.clone();
///
/// // Expensive (heap allocated string).
/// let a = Time::from(3234815_u64);
/// assert!(a == "1 month, 6 days, 23 hours, 59 minutes, 59 seconds");
/// let b = a.clone();
/// ```
///
/// The actual string used internally is not a [`String`](https://doc.rust-lang.org/std/string/struct.String.html),
/// but a [`CompactString`](https://docs.rs/compact_str) so that any string 24 bytes (12 bytes on 32-bit) or less are _stack_ allocated instead of _heap_ allocated.
///
/// The documentation will still refer to the inner string as a `String`. Anything returned will also be a `String`.
///
/// ## Math
/// These operators are overloaded. They will always output a new [`Self`]:
/// - `Add +`
/// - `Sub -`
/// - `Div /`
/// - `Mul *`
/// - `Rem %`
///
/// They can either be:
/// - Combined with another [`Self`]: `Time::from(1) + Time::from(1)`
/// - Or with the inner number itself: `Time::from(1) + 1`
///
/// They also have the same `panic!()` behavior on overflow as the normal ones, because internally,
/// it is just calling `.inner() $OPERATOR $NUMBER`.
///
/// ```rust
/// # use readable::*;
/// assert!(Time::from(10_u64) + 10 == Time::from(20_u64));
/// assert!(Time::from(10_u64) - 10 == Time::from(0_u64));
/// assert!(Time::from(10_u64) / 10 == Time::from(1_u64));
/// assert!(Time::from(10_u64) * 10 == Time::from(100_u64));
/// assert!(Time::from(10_u64) % 10 == Time::from(0_u64));
/// ```
/// Overflow example:
/// ```rust,should_panic
/// # use readable::*;
/// let n = Time::from(u64::MAX) + u64::MAX;
/// ```
///
/// ## Examples
/// ```rust
/// # use readable::Time;
/// assert!(Time::from(0_u64)        == "0 seconds");
/// assert!(Time::from(1_u64)        == "1 second");
/// assert!(Time::from(2_u64)        == "2 seconds");
/// assert!(Time::from(59_u64)       == "59 seconds");
/// assert!(Time::from(60_u64)       == "1 minute");
/// assert!(Time::from(61_u64)       == "1 minute, 1 second");
/// assert!(Time::from(62_u64)       == "1 minute, 2 seconds");
/// assert!(Time::from(120_u64)      == "2 minutes");
/// assert!(Time::from(121_u64)      == "2 minutes, 1 second");
/// assert!(Time::from(122_u64)      == "2 minutes, 2 seconds");
/// assert!(Time::from(179_u64)      == "2 minutes, 59 seconds");
/// assert!(Time::from(3599_u64)     == "59 minutes, 59 seconds");
/// assert!(Time::from(3600_u64)     == "1 hour");
/// assert!(Time::from(3601_u64)     == "1 hour, 1 second");
/// assert!(Time::from(3602_u64)     == "1 hour, 2 seconds");
/// assert!(Time::from(3660_u64)     == "1 hour, 1 minute");
/// assert!(Time::from(3720_u64)     == "1 hour, 2 minutes");
/// assert!(Time::from(86399_u64)    == "23 hours, 59 minutes, 59 seconds");
/// assert!(Time::from(86400_u64)    == "1 day");
/// assert!(Time::from(86401_u64)    == "1 day, 1 second");
/// assert!(Time::from(86402_u64)    == "1 day, 2 seconds");
/// assert!(Time::from(86460_u64)    == "1 day, 1 minute");
/// assert!(Time::from(86520_u64)    == "1 day, 2 minutes");
/// assert!(Time::from(90000_u64)    == "1 day, 1 hour");
/// assert!(Time::from(93600_u64)    == "1 day, 2 hours");
/// assert!(Time::from(604799_u64)   == "6 days, 23 hours, 59 minutes, 59 seconds");
/// assert!(Time::from(604800_u64)   == "7 days");
/// assert!(Time::from(2630016_u64)  == "1 month");
/// assert!(Time::from(3234815_u64)  == "1 month, 6 days, 23 hours, 59 minutes, 59 seconds");
/// assert!(Time::from(5260032_u64)  == "2 months");
/// assert!(Time::from(31557600_u64) == "1 year");
/// assert!(Time::from(63115200_u64) == "2 years");
/// assert_eq!(
///     Time::from(u64::MAX),
///     "584542046090 years, 7 months, 15 days, 17 hours, 5 minutes, 3 seconds",
/// );
/// ```
/// # Credit
/// This code is forked from `https://docs.rs/humantime`, edited to remove sub-second time, change spacing and some words.
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "bincode", derive(bincode::Encode, bincode::Decode))]
#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
pub struct Time(u64, #[cfg_attr(feature = "bincode", bincode(with_serde))] CompactString);

// Implementation Macro.
macro_rules! impl_number {
	($number:ty) => {
		impl From<$number> for Time {
			#[inline]
			fn from(number: $number) -> Self {
				Self::from(number as u64)
			}
		}
	}
}

impl_math!(Time, u64);
impl_traits!(Time, u64);
impl_number!(u8);
impl_number!(u16);
impl_number!(u32);
impl_number!(usize);

impl Time {
	impl_common!(u64);
	impl_not_const!();
	impl_usize!();

	#[inline]
	fn plural(string: &mut CompactString, started: &mut bool, name: &str, value: u64) {
		if value > 0 {
			if *started {
				string.push_str(", ");
			}
			let _ = write!(string, "{} {}", value, name);
			if value > 1 {
				string.push('s');
			}
			*started = true;
		}
	}

	#[inline]
	/// ```rust
	/// # use readable::Time;
	/// assert!(Time::zero() == 0_u64);
	/// assert!(Time::zero() == "0 seconds");
	/// ```
	pub const fn zero() -> Self {
		Self(0, CompactString::new_inline("0 seconds"))
	}

	#[inline]
	/// ```rust
	/// # use readable::Time;
	/// assert!(Time::second() == 1_u64);
	/// assert!(Time::second() == "1 second");
	/// ```
	pub const fn second() -> Self {
		Self(1, CompactString::new_inline("1 second"))
	}

	#[inline]
	/// ```rust
	/// # use readable::Time;
	/// assert!(Time::minute() == 60_u64);
	/// assert!(Time::minute() == "1 minute");
	/// ```
	pub const fn minute() -> Self {
		Self(60, CompactString::new_inline("1 minute"))
	}

	#[inline]
	/// ```rust
	/// # use readable::Time;
	/// assert!(Time::hour() == 3600_u64);
	/// assert!(Time::hour() == "1 hour");
	/// ```
	pub const fn hour() -> Self {
		Self(3600, CompactString::new_inline("1 hour"))
	}

	#[inline]
	/// ```rust
	/// # use readable::Time;
	/// assert!(Time::day() == 86400_u64);
	/// assert!(Time::day() == "1 day");
	/// ```
	pub const fn day() -> Self {
		Self(86400, CompactString::new_inline("1 day"))
	}

	#[inline]
	/// ```rust
	/// # use readable::Time;
	/// assert!(Time::month() == 2630016_u64);
	/// assert!(Time::month() == "1 month");
	/// ```
	pub const fn month() -> Self {
		Self(2630016, CompactString::new_inline("1 month"))
	}

	#[inline]
	/// ```rust
	/// # use readable::Time;
	/// assert!(Time::year() == 31557600_u64);
	/// assert!(Time::year() == "1 year");
	/// ```
	pub const fn year() -> Self {
		Self(31557600, CompactString::new_inline("1 year"))
	}
}

impl From<std::time::Duration> for Time {
	fn from(duration: std::time::Duration) -> Self {
		Self::from(duration.as_secs())
	}
}

impl From<&std::time::Duration> for Time {
	fn from(duration: &std::time::Duration) -> Self {
		Self::from(duration.as_secs())
	}
}

impl From<std::time::Instant> for Time {
	fn from(instant: std::time::Instant) -> Self {
		Self::from(instant.elapsed().as_secs())
	}
}

impl From<&std::time::Instant> for Time {
	fn from(instant: &std::time::Instant) -> Self {
		Self::from(instant.elapsed().as_secs())
	}
}

impl From<u64> for Time {
	fn from(secs: u64) -> Self {
		#[cfg(feature = "inline_time")]
		if secs <= 3660 {
			// SAFETY:
			// Cast `u64` to `u16` is safe because it's under 65_535.
			return Self(secs, CompactString::new_inline(readable_inlined_time::inlined(secs as u16)))
		}

		if secs == 0 {
			return Self::zero()
		}

		let years = secs / 31_557_600;  // 365.25d
		let ydays = secs % 31_557_600;
		let months = ydays / 2_630_016;  // 30.44d
		let mdays = ydays % 2_630_016;
		let days = mdays / 86400;
		let day_secs = mdays % 86400;
		let hours = day_secs / 3600;
		let minutes = day_secs % 3600 / 60;
		let seconds = day_secs % 60;

		let started = &mut false;
		let mut string = CompactString::with_capacity(8);
		Self::plural(&mut string, started, "year", years);
		Self::plural(&mut string, started, "month", months);
		Self::plural(&mut string, started, "day", days);
		Self::plural(&mut string, started, "hour", hours);
		Self::plural(&mut string, started, "minute", minutes);
		Self::plural(&mut string, started, "second", seconds);

		Self(secs, string)
	}
}

//---------------------------------------------------------------------------------------------------- TESTS
#[cfg(test)]
mod tests {
	
}
